lib/curl_sha512_256.c

```diff
@@ -0,0 +1,588 @@
+ * ** under curl license.                                                ** *
+ * ** This is a minimal version with minimal optimisations. Performance  ** *
+ * ** can be significantly improved. Big-endian store and load macros    ** *
+ * ** are obvious targets for optimisation.                              ** */
+
+#if !defined(MHDX_UINT64) && defined(SIZEOF_LONG)
+#if SIZEOF_LONG == 8
+#define MHDX_UINT64 unsigned long
+#define MHDX_UINT64_C(val) val##UL
+#endif /* SIZEOF_LONG == 8 */
+#endif /* !MHDX_UINT64 && SIZEOF_LONG */
+
+#if !defined(MHDX_UINT64) && defined(HAVE_LONGLONG)
+#define MHDX_UINT64 unsigned long long
+#define MHDX_UINT64_C(val) val##ULL
+#endif /* !MHDX_UINT64 && SIZEOF_LONG */
+
+#if !defined(MHDX_UINT64)
+#define MHDX_UINT64 unsigned __int64
+#define MHDX_UINT64_C(val) ((MHDX_UINT64)val)
+#endif /* !MHDX_UINT64 && SIZEOF_LONG */
+
+#ifndef MHDX_UINT64_TRIM
+/* Trim possibly larger value to 64 bits.
+   Assuming that no type is larger than 64 bits. */
+#define MHDX_UINT64_TRIM(x) ((MHDX_UINT64)(x))
+#endif /* !MHDX_UINT64_TRIM */
+
+/* Bits manipulation macros and functions.
+   Can be moved to other headers to reuse. */
+
+#define MHDX_GET_64BIT_BE(ptr)                                \
+  MHDX_UINT64_TRIM(                                           \
+    ((MHDX_UINT64)(((const unsigned char*)(ptr))[0]) << 56) | \
+    ((MHDX_UINT64)(((const unsigned char*)(ptr))[1]) << 48) | \
+    ((MHDX_UINT64)(((const unsigned char*)(ptr))[2]) << 40) | \
+    ((MHDX_UINT64)(((const unsigned char*)(ptr))[3]) << 32) | \
+    ((MHDX_UINT64)(((const unsigned char*)(ptr))[4]) << 24) | \
+    ((MHDX_UINT64)(((const unsigned char*)(ptr))[5]) << 16) | \
+    ((MHDX_UINT64)(((const unsigned char*)(ptr))[6]) << 8)  | \
+    (MHDX_UINT64)(((const unsigned char*)(ptr))[7]) )
+
+#define MHDX_PUT_64BIT_BE(ptr,val) do {                                   \
+  ((unsigned char*)(ptr))[7]=(unsigned char)((MHDX_UINT64)(val));         \
+  ((unsigned char*)(ptr))[6]=(unsigned char)(((MHDX_UINT64)(val)) >> 8);  \
+  ((unsigned char*)(ptr))[5]=(unsigned char)(((MHDX_UINT64)(val)) >> 16); \
+  ((unsigned char*)(ptr))[4]=(unsigned char)(((MHDX_UINT64)(val)) >> 24); \
+  ((unsigned char*)(ptr))[3]=(unsigned char)(((MHDX_UINT64)(val)) >> 32); \
+  ((unsigned char*)(ptr))[2]=(unsigned char)(((MHDX_UINT64)(val)) >> 40); \
+  ((unsigned char*)(ptr))[1]=(unsigned char)(((MHDX_UINT64)(val)) >> 48); \
+  ((unsigned char*)(ptr))[0]=(unsigned char)(((MHDX_UINT64)(val)) >> 56); \
+} while(0)
+
+/* Defined as a function. The macro version may duplicate the binary code
+ * size as each argument is used twice, so if any calculation is used
+ * as an argument, the calculation could be done twice. */
+static inline MHDX_UINT64
+MHDx_rotr64(MHDX_UINT64 value, unsigned int bits)
+{
+  bits %= 64;
+  if(0 == bits)
+    return value;
+  /* Defined in a form which modern compiler could optimise. */
+  return MHDX_UINT64_TRIM((value >> bits) | (value << (64 - bits)));
```

Comment:
This does not need trimming, it can't be larger than 64 bit.

Reply:
Initial idea was to support types larger that 64 bits.
I current form it's not needed. I'll correct.